### ISSUE TITLE
[css-properties-values-api-1] What is the computed value of a <transform-list> in the middle of a layout dependent matrix decomposed interpolation?

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#2854]
_From @alancutter on July 6, 2017 5:4_

The [`<transform-list>` syntax](https://drafts.css-houdini.org/css-properties-values-api/#supported-syntax-strings) enables custom properties to animate like the `transform` property.
One quirk of `transform` is that the [serialisation procedure](https://drafts.csswg.org/css-transforms-1/#serialization-of-the-computed-value) requires layout information in order to turn percentages of the element's box dimensions into a matrix.

Unfortunately this procedure requires layout information to turn an animation between `translateX(100%)` and `rotate(45deg)` into a single matrix. There is currently no way of representing such a value in CSS independent of what the layout is. You would need something like `blend(translateX(100%), rotate(45deg), 0.5)`.

Here's a more concrete example:
```html
<!DOCTYPE html>
<style>
#target {
  /* Animation is paused at halfway. */
  animation: x 1s -0.5s linear paused;
  --y: var(--x);
}
@keyframes x {
  from { --x: translateX(100%); }
  to { --x: rotate(45deg); }
}
</style>

<div id="target"></div>

<script>
CSS.registerProperty({
  name: '--x',
  syntax: '<transform-list>',
  initialValue: 'none',
});

// What should the computed value of --y be?
console.log(getComputedStyle(target).getPropertyValue('--y'));
</script>
```

Should the serialisation of the computed value of `<transform-list>` custom properties be the same as `transform` and depend on layout computations?
Should this layout computation dependency "infect" other custom properties that reference it via `var()`?
What happens when we try to serialise a `<transform-list>` custom property inside a layout worklet in the middle of layout computation?

I don't think depending on layout computation is a good idea here.
I recommend using `<transform-list-2> := <transform-list> | blend(<transform-list-2>, <transform-list-2>, <number>)` instead (but with a better name) and to serialise it as the original transform functions instead of as a matrix that incorporates layout computations.

_Copied from original issue: w3c/css-houdini-drafts#425_

**Comment by tabatkins** [comment]
_From @shans on October 19, 2017 0:6_

We definitely can't depend on layout to produce computed styles. I think something like your blend suggestion makes sense. Flagging this for discussion with the Houdini TF.

**Comment by tabatkins** [comment]
_From @css-meeting-bot on November 9, 2017 18:42_

The Working Group just discussed `What is the computed value of a <transform-list> in the middle of a layout dependent matrix decomposed interpolation?`, and agreed to the following resolutions:

* `RESOLVED: Add generic interpolation function. Name TBD. For Units and Values spec.`

<details><summary>The full IRC log of that discussion</summary>
&lt;TabAtkins> Topic: What is the computed value of a &lt;transform-list> in the middle of a layout dependent matrix decomposed interpolation?<br>
&lt;TabAtkins> GitHub: https://github.com/w3c/css-houdini-drafts/issues/425<br>
&lt;nainar> Scribenick: nainar<br>
&lt;nainar> TabAtkins: Problem - you have a rtanstition between translate 30% and rotation 30 deg. The 30% is less dependant on lyaout info.<br>
&lt;nainar> TabAtkins: you are in the middle of transition - due to incomaptiblilty you get matrix<br>
&lt;nainar> TabAtkins: What should CSSOM return for Computed value?<br>
&lt;nainar> TabAtkins: string based computd vaue can return layout dependant, but typed om cant<br>
&lt;nainar> dbaron: interpolatematrix in gecko internally. This is important for intepolation rules for transform Needs to interpolate between partial inter-polation results. So that you can change in the middle3of transtion<br>
&lt;nainar> dbaron: There is other rationale for interpolation this.<br>
&lt;nainar> dbaron: Thought we agrred to add this to transform spec?<br>
&lt;nainar> TabAtkins: Happy to add it but not there currently<br>
&lt;nainar> surma: why cant we interpolate on % value<br>
&lt;nainar> dbaron: you cant represent computed value without layout<br>
&lt;nainar> TabAtkins: GCS does used vaoue for this. We are trying to return ComputedValue which isnt representable as a list<br>
&lt;nainar> surma: Sounds like you need to solve geenrically.<br>
&lt;nainar> TabAtkins: Add interpolate function that interpolates between two values. Like cross fade<br>
&lt;nainar> TabAtkins: Proposal name is blend<br>
&lt;nainar> dbaron: Sounds too imge-y<br>
&lt;nainar> mwoodrow: Happy to accept concept not name<br>
&lt;nainar> dholbert: You could simplify calc in many cases.<br>
&lt;dholbert> s/calc/to calc/<br>
&lt;nainar> TabAtkins: computed value from 10 px to 10 px should be a px value not the function<br>
&lt;nainar> RESOLVED: Add generic interpolation function. Name TBD. For Units and Values spec.<br>
&lt;dbaron> s/Units and Values/Values and Units/<br>
</details>

**Comment by fantasai** [comment]
@tabatkins Edits checked in, please look them over and close out the issue if they're acceptable to resolve this issue?

**Comment by tabatkins** [comment]
Yeah, this'll need some edits in Transforms to specify that it's used, but `mix()` is indeed sufficient for the purposes.
